cpp/src/io/csv/datetime.cuh

```diff
@@ -382,58 +288,62 @@ __inline__ __device__ T parse_optional_integer(char const** begin, char const* e
  * @param end Pointer to the first element after the string
- * @return The parsed duration
+ * @return The parsed duration in `duration_type`
  */
-template <typename T>
-__inline__ __device__ int64_t to_time_delta(char const* begin, char const* end)
+template <typename duration_type>
+__inline__ __device__ duration_type to_duration(char const* begin, char const* end)
 {
+  using cuda::std::chrono::duration_cast;
+
   // %d days [+]%H:%M:%S.n => %d days, %d days [+]%H:%M:%S,  %H:%M:%S.n, %H:%M:%S, %value.
   constexpr char sep = ':';
 
-  int32_t days{0};
-  int8_t hour{0};
   // single pass to parse days, hour, minute, seconds, nanosecond
   auto cur         = begin;
   auto const value = parse_integer<int32_t>(&cur, end);
   cur              = skip_spaces(cur, end);
-  if (std::is_same_v<T, cudf::duration_D> || cur >= end) {  // %value
-    return value;
+  if (std::is_same_v<duration_type, cudf::duration_D> || cur >= end) {
+    return duration_type{static_cast<typename duration_type::rep>(value)};
   }
+
   // " days [+]"
   auto const after_days_sep     = skip_if_starts_with(cur, end, "days");
   auto const has_days_seperator = (after_days_sep != cur);
   cur                           = skip_spaces(after_days_sep, end);
   cur += (*cur == '+');
+
+  duration_D d_d{0};
+  duration_h d_h{0};
   if (has_days_seperator) {
-    days = value;
-    hour = parse_integer<int8_t>(&cur, end);
+    d_d = duration_D{value};
+    d_h = duration_h{parse_integer<int32_t>(&cur, end)};
   } else {
-    hour = value;
+    d_h = duration_h{value};
   }
 
-  auto const minute = parse_optional_integer<int8_t>(&cur, end, sep);
-  auto const second = parse_optional_integer<int8_t>(&cur, end, sep);
-
-  int nanosecond = 0;
-  if (std::is_same_v<T, cudf::duration_s>) {
-    return ((days * 24L + hour) * 60L + minute) * 60L + second;
-  } else if (*cur == '.') {  //.n
-    auto const start_subsecond        = ++cur;
-    nanosecond                        = parse_integer<int>(&cur, end);
-    int8_t const num_digits           = min(9L, cur - start_subsecond);
-    constexpr int64_t powers_of_ten[] = {
-      1L, 10L, 100L, 1000L, 10000L, 100000L, 1000000L, 10000000L, 100000000L, 1000000000L};
-    nanosecond *= powers_of_ten[9 - num_digits];
+  duration_m d_m{parse_optional_integer<int32_t>(&cur, end, sep)};
+  duration_s d_s{parse_optional_integer<int64_t>(&cur, end, sep)};
+
+  // Convert all durations to the given type
+  auto output_d = duration_cast<duration_type>(d_d + d_h + d_m + d_s);
+
+  if constexpr (std::is_same_v<duration_type, cudf::duration_s>) { return output_d; }
+
+  // Duration type not allowed due to the use of `*` operator
+  int64_t nanosecond = 0L;
+  if (*cur == '.') {  //.n
+    auto const start_subsecond = ++cur;
+    nanosecond                 = parse_integer<int64_t>(&cur, end);
+    int8_t const num_digits    = min(9L, cur - start_subsecond);
+    nanosecond *= powers_of_ten[9 - num_digits];  // Duration has no * operator
```

Comment:
I'm confused what's going on here. Why are we scaling the `nanosecond` value depending on the number of digits? This function also needs some love to describe what the allow formats are and what they mean.

Reply:
Oh, the name is misleading. If there are more than 9 digits after the decimal point, we truncate raw time digits to nanoseconds. Otherwise, we scale the raw time.

Reply:
Should be OK now.

Reply:
Okay, I see what's going on here. If I have a string that's like `"13:01:27.1379"`, then you want to take `0.1379` seconds and convert it into a count of nanoseconds (`137900000`). So you parse the integer value `1379` and multiply it by 10e5 to get the nanoseconds. 

If the string was instead `"13:01:27.1379000001"`, you want to truncate that to ``"13:01:27.137900000"`. But I don't think the code as-is will do that. `parse_integer<int64_t>` would return `1379000001` which will end up doing `1379000001 * powers_of_ten[0]`, which is `1379000001`, and you'll get the wrong count of nanoseconds.

Reply:
Also, to make this easier to understand what's going on, I would rename the return value of `parse_integer` here to be something like `fraction` or `decimal`.

Reply:
Cool, so this is actually like a fixed point number rescaling. So something like this;
```
auto const ns = (*cur != '.') ? duration_ns{0} : [&](){
   auto const start_subsecond = ++cur;
   auto const unscaled_subseconds = parse_integer<int64_t>(&cur, end);
   auto const scale = 9 - count_digits(subseconds); // max of 9 digits in nanoseconds
   auto const rescaled  = decimal64{unscaled_subseconds, scale_type{scale}}.rescaled(scale_type{0});
   return duration_ns{rescaled.value()};
}();
```

Reply:
With https://github.com/rapidsai/cudf/issues/9346, this fixed_point implementation should be just as fast as the original.